File: 2-hard/ascii_decryption/ascii_decryption.php
<?php

function decrypt_message($msg, $wordLength, $lastChar)
{
    $lastCharAscii = ord($lastChar);
    $spaceAscii = ord(" ");
    $difference = $lastCharAscii - $spaceAscii;
    $msg = explode(" ", $msg);

    $repeated = array();
    for ($i = 0; $i < count($msg) - 1; $i++) {
        $msg[$i] = intval($msg[$i]);
        if ($msg[$i] - $msg[$i + 1] == $difference && $msg[$i - $wordLength] == $msg[$i + 1]) {
            $repeated[] = $i;
        }
    }

    if (count($repeated) == 2) {
        $n = $lastCharAscii - $msg[$repeated[0]];
        for ($i = 0; $i < count($msg); $i++) {
            $msg[$i] = chr($msg[$i] + $n);
        }

        return join("", $msg);
    }

}

$fileName = $argv[1];
$content = file_get_contents($fileName);
list($wordLength, $lastChar, $msg) = explode(" | ", $content);
echo decrypt_message($msg, intval($wordLength), $lastChar)."\n";

File: 2-hard/ascii_decryption/ascii_decryption.rb
def decrypt_message(msg, word_length, last_char)
  last_char_ascii = last_char.ord
  space_ascii = " ".ord
  difference = last_char_ascii - space_ascii

  repeated = []
  for i in 0..msg.count - 2
    if msg[i] - msg[i + 1] == difference && msg[i - word_length] == msg[i + 1]
      repeated << i
    end
  end
  if repeated.count == 2
    n = last_char_ascii - msg[repeated.first]
    return msg.map {|ascii| (ascii + n).chr }.join

  end
end


content = File.read(ARGV[0])
word_length, last_char, encrypted_message = content.split(" | ")
encrypted_message = encrypted_message.split(" ").map {|char| char.to_i}
last_char_ascii = last_char.ord
space_ascii = " ".ord
puts decrypt_message(encrypted_message, word_length.to_i, last_char)


